cpp/src/arrow/python/udf.cc

```diff
@@ -0,0 +1,131 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/python/udf.h"
+#include "arrow/compute/function.h"
+#include "arrow/python/common.h"
+
+namespace arrow {
+
+namespace py {
+
+namespace {
+Status CheckOutputType(const DataType& expected, const DataType& actual) {
+  if (!expected.Equals(actual)) {
+    return Status::TypeError("Expected output type, ", expected.ToString(),
+                             ", but function returned type ", actual.ToString());
+  }
+  return Status::OK();
+}
+
+struct PythonUdf {
+  ScalarUdfWrapperCallback cb;
+  std::shared_ptr<OwnedRefNoGIL> function;
+  compute::OutputType output_type;
+
+  // function needs to be destroyed at process exit
+  // and Python may no longer be initialized.
+  ~PythonUdf() {
+    if (_Py_IsFinalizing()) {
+      function->detach();
+    }
+  }
+
+  Status operator()(compute::KernelContext* ctx, const compute::ExecBatch& batch,
+                    Datum* out) {
+    return SafeCallIntoPython([=]() -> Status { return Execute(ctx, batch, out); });
```

Comment:
```suggestion
    return SafeCallIntoPython([&]() -> Status { return Execute(ctx, batch, out); });
```
Minor nit but this will force a copy of a `batch` if you have `=` and we can avoid it with `&`.  [Example](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIAOykrgAyeAyYAHI%2BAEaYxCAAzABspAAOqAqETgwe3r4BaRlZAqHhUSyx8cm2mPaOAkIETMQEuT5%2BgXaYDtmNzQSlkTFxiSkKTS1t%2BZ0TA2FDFSPJAJS2qF7EyOwc5glhyN5YANQmCW5sLCQAnqfYJhoAgrv7h5gnZ07jxJist/dPjwOTAUCiOABViExkABrTDobCCQhXE7%2BKyPI4Yo6pLzRWh4ZAgf7/THgyEwuEIxwEK4QZYotEPEkk8boEAodYEd5uU5uE5mMwQqGw%2BGI6nmMxcnlHFls1y0U4MkkmfwAEWJmMF5JFVJpaAY41JQopopuZiSR1QBAQcTpysVTIxMvZXk5PKl4o8qVNErdZ2lBFZIDlCvVGOVavRGrJwspSIgmpjJvMSWTFqtNvpoaZTrQLslfvFAFlUAA3TDi/O8p3BhL2sOqrMJ4061OoVJxJhEYinFUQPUGpvapGt9PEW2orPMgNs3Ous7u/memnAzLABjLCu%2BqvToMMdDy2uTzHfAgbBhHABUVrwChDkfrEcZUaNQ7FZot7chXZ78ejzeHZojtaY6Zve2Y7rOlZ8mYxZlhAK54GuG78lB1Z7gedZMieZ6Xtet6HmB4ZEo8REEQCDxhJyLBMGEtKgU%2BGKDrG1JHIwLZkVOgaQVu0FPChPFofud4MUcTAuqgRwGBU6BMFw7wqicACsVgJGqim9uOmEkthxDnmxw4cZipGYTmHJQeKVj8fOfqCRhWZiUQkmsNEMk%2BqpSmWSm6l0XaR4YjpelJoZD7CZxM5mTxHqbtZ26BjWmEBUcGjCeGHCrLQnCKbwfgcFopCoJw3KWNY0rrJsby7DwpAEJoaWrNCiQJAAdAkrVte17UpBlHCSNltX5ZwvAKCAGjVbVqxwLASBoCwqR0HE5CUDNc30PEwBcFwCR8HQBBxMNEDRP10RhM0Vz9TNbCCAA8gwtBnblvBYNRRjiA9pD4N8PRlsNb2YKo3QutseWUbU/V4tEkLEFcHhYP1BDEHgLCcDwqxUAYwAKAAangmAAO5Xe2OVVfwggiGI7BSDIgiKCo6hvboXD6IYxjWNY%2Bh4NEw2QKsbb1PqnAALRXQkRwCyyPamMVlgbUNtTdHzLh7lMfiMyE8zlJUejpJkfPK1rRR84MGsjIzXQ9A0sx66bcvmwwfQtEbwzxKblueO0ejjP0juLM7qwKGVWx6PDmDbCj%2BiZX1b0FRwqgABxJALSSSEcwDIMgRwbS1RwQEVlk2EcuCECQfIJIzRweLN83ECXXDLLwNUPcs9WNS1HVt61XWcL1pA5Xl0dDSNY2N6Qk2ICAHLYgQi0QHCRcI6yjMk8IojiJTS802o/UM6QuOQqkyNpeHHBZT3/XR1dLqTxaVBHHHCdJynacZ1wWc56glerTXddD1oTekA1rVW7tw6kfbuvdeD91sIPBuv90qcDMJHPug0f51VIGWYgmRnCSCAA%3D%3D%3D)

The rules for lambda captures are a bit odd :)

Reply:
Thank you for cathing this and for the nice explanation. I am updating it.